src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+    shared_ptr<PatternDatabase> pdb =
+        make_shared<PatternDatabase>(concrete_task_proxy, pattern);
+    shared_ptr<AbstractTask> projected_task =
+        extra_tasks::build_projected_task(concrete_task, move(pattern));
+    TaskProxy projected_task_proxy(*projected_task);
+
+    bool unsolvable = false;
+    vector<vector<OperatorID>> plan;
+    size_t initial_state_index = pdb->hash_index_of_projected_state(
+        projected_task_proxy.get_initial_state());
+    int init_goal_dist = pdb->get_value_for_hash_index(initial_state_index);
+    if (init_goal_dist == numeric_limits<int>::max()) {
+        unsolvable = true;
+        if (verbosity >= utils::Verbosity::VERBOSE) {
+            utils::g_log << "PDB with pattern " << pdb->get_pattern()
+                         << " is unsolvable" << endl;
+        }
+    } else {
+        if (verbosity >= utils::Verbosity::VERBOSE) {
+            utils::g_log << "Computing plan for PDB with pattern "
+                         << pdb->get_pattern() << endl;
+        }
+
+        plan = steepest_ascent_enforced_hill_climbing(
+            projected_task_proxy, rng, *pdb, compute_wildcard_plan, verbosity);
+
+        // Convert operator IDs of the abstract in the concrete task.
+        for (vector<OperatorID> &plan_step : plan) {
+            vector<OperatorID> concrete_plan_step;
+            concrete_plan_step.reserve(plan_step.size());
+            for (OperatorID abs_op_id : plan_step) {
+                concrete_plan_step.push_back(
+                    projected_task_proxy.get_operators()[abs_op_id].get_ancestor_operator_id(concrete_task.get()));
+            }
+            plan_step.swap(concrete_plan_step);
+        }
+    }
+
+    return utils::make_unique_ptr<Projection>(
+        move(pdb), move(plan), unsolvable);
+}
+
+void CEGAR::print_collection() const {
+    utils::g_log << "[";
+    for (size_t i = 0; i < projection_collection.size(); ++i) {
+        const unique_ptr<Projection> &projection = projection_collection[i];
+        if (projection) {
+            utils::g_log << projection->get_pattern();
+            if (i != projection_collection.size() - 1) {
+                utils::g_log << ", ";
+            }
+        }
+    }
+    utils::g_log << "]" << endl;
+}
+
+void CEGAR::compute_initial_collection() {
+    assert(!goals.empty());
+    for (const FactPair &goal : goals) {
+        add_pattern_for_var(goal.var);
+    }
+
+    if (verbosity >= utils::Verbosity::VERBOSE) {
+        utils::g_log << "initial collection: ";
+        print_collection();
+        utils::g_log << endl;
+    }
+}
+
+bool CEGAR::time_limit_reached(
+    const utils::CountdownTimer &timer) const {
+    if (timer.is_expired()) {
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "time limit reached." << endl;
+        }
+        return true;
+    }
+    return false;
+}
+
+/*
+  TODO: this is a duplicate of State::get_unregistered_successor.
```

Comment:
I think it is a very important assertion, though, which I would like to have in my regular code. Therefore I'm not sure what to do about it and would wait/ask for Malte's opinion.

Reply:
Which one is the added assertion? Line 115 seems redundant because it is implied by line 125, so I assume line 125 is a relevant one. Is there a reason why we want to assert this for every successors rather than -- like we do for other heuristics -- once, in a larger context? If we absolutely need this assertion here and it is the only change compared to the original `get_unregistered_successor`, why do we need to copy the code rather than have a wrapper function that consists just of a call to the original `get_unregistered_successor` and the assertion?

Reply:
Sorry for the confusion, but it is about neither of these assertions, but about one the original method has, which I copy here:
```
State State::get_unregistered_successor(const OperatorProxy &op) const {
    assert(!op.is_axiom());
    assert(task_properties::is_applicable(op, *this));
    assert(values);
    vector<int> new_values = get_unpacked_values();

    for (EffectProxy effect : op.get_effects()) {
        if (does_fire(effect, *this)) {
            FactPair effect_fact = effect.get_fact().get_pair();
            new_values[effect_fact.var] = effect_fact.value;
        }
    }

    if (task->get_num_axioms() > 0) {
        AxiomEvaluator &axiom_evaluator = g_axiom_evaluators[TaskProxy(*task)];
        axiom_evaluator.evaluate(new_values);
    }
    return State(*task, move(new_values));
}
```
The is_applicable assertion doesn't hold in our case. I further replaced the code that does axiom evaluation by the assertion on line 125. The assertion on line 115 is also present in the original code. 

The only difference we would really need is to not assert that the operator is applicable.

Reply:
I see. Yes, I agree this is a valuable assertion. This perhaps needs to be considered in a wider context then. If it doesn't already, I think it would be good for the version in `cegar.cc` to explain how it differs from the original function and why this is desirable. (We can discuss this later.)

Reply:
I assume you have reasons not to change this to something along the lines of
``` 
State State::get_unregistered_successor(const OperatorProxy &op) const {
    assert(task_properties::is_applicable(op, *this));
    return get_unregistered_successor_ignoring_preconditions(op);
}

State State::get_unregistered_successor_ignoring_preconditions(const OperatorProxy &op) const {
    // original function without the assertion
}
```

Reply:
I don't know. Malte, what wider context did you mean?

Reply:
If we ignore preconditions, it is not really a successor, so it is a somewhat weird concept to be promoted to the (very prominent) public interface of the state class. If we don't have such a function, we have potentially useful code invariants that get lost if we add it (only reachable states can be generated).

By wider context I meant other solutions to the same problem, or ways to avoid needing this in the first place. To judge this I would need to know more about where this is used and why.

An example of another solution to the same problem could be a task adapter that ignores preconditions. But I also wonder why we want this function.

Reply:
Ok. I don't think we need anything of this.

Instead, I'd suggest to not use `State` at all here but unpacked values, i.e., `vector<int>`. The function in question would just be called `apply_operator`, not check applicability and update the vector in-place, saving the copy. 

At the end, we would only once need to construct a `State` object from the vector because we want to check if the resulting state is a goal state.